File: C++/022_Generate Parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector <string> res;
        char *s = new char[(n << 1) + 1];
        s[n << 1] = '\0';
        res.clear();
        gen(res, s, 0, n, n);
        return res;
    }
    
    void gen(vector <string> &v, char *s, int x, int l, int r) {
        if (0 == l && 0 == r) {
            v.push_back(s);
            return ;
        }
        if (l) {
            s[x] = '(';
            gen(v, s, x + 1, l - 1, r);
        }
        if (r > l) {
            s[x] = ')';
            gen(v, s, x + 1, l, r - 1);
        }
    }
};

File: Python/022_Generate Parentheses.py
class Solution(object):
    def __init__(self):
        self.res = []

    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        self._dfs('', 0, 0, n)
        return self.res

    def _dfs(self, s, l, r, n):
        if l >= n and r >= n:
            self.res.append(s)
            return
        if l < n:
            self._dfs(s + '(', l + 1, r, n)
        if l > r:
            self._dfs(s + ')', l, r + 1, n)